### DISCUSSION TITLE
Adding tools for DG methods to amrex code-base?

### DISCUSSION THREAD

**Discussion by dunhamsj** [AMReX-Codes/amrex#4178] [General]
Hello, I've made additions to my [fork](https://github.com/dunhamsj/amrex) of amrex that allow a user to call FillPatch and AverageDown on a MultiFab containing a nodal DG solution for each field. Would it be possible to merge these features into the main code-base of amrex? They have not been tested on GPUs, but they pass all of the github actions.

**Comment by WeiqunZhang** [comment]
Could you explain what's "a MultiFab containing a nodal DG solution for each field" and what you need to do in FillPatch and AverageDown?

**Reply by dunhamsj** [reply]
Yes, sorry, that was vague. What I mean is that, in a case where you evolve the density, rho, and the energy density, e, with four degrees of freedom per field per element, the fourth dimension of the multifab would be ( rho_{1}, rho_{2}, rho_{3}, rho_{4}, e_{1}, e_{2}, e_{3}, e_{4} ). The FillPatch and AverageDown routines involve a matrix-vector multiply, where the vector contains the degrees of freedom for a particular field and the matrix is a projection matrix, which has to be supplied by the user. The elements of the projection matrix are defined as integrals over reference elements of the product of the basis functions of the DG solution. I should also mention this code is not able to handle p-refinement.

**Reply by WeiqunZhang** [reply]
Could you post links to your modified FillPatch and AverageDown functions? I could try to find it myself in your fork. But it might save me some time, if you could point me to them.

**Reply by dunhamsj** [reply]
Sure. There are several files that involve them and I'm not sure exactly which ones you want.

The actual interpolation for `FillPatch` for 3D are lines 35 and 128 [here](https://github.com/dunhamsj/amrex/blob/development/Src/AmrCore/AMReX_Interp_3D_C.H). Similar routines exist for 1D and 2D.

The actual interpolation for `AverageDown` for 3D are lines 456, 519 [here](https://github.com/dunhamsj/amrex/blob/development/Src/Base/AMReX_MultiFabUtil_3D_C.H). Similar routines exist for 1D and 2D.

There are two routines: one that interpolates in such a way as to conserve the cell average, and another that just interpolates pointwise.

**Other places they are referenced**

**FillPatch**
Lines 278 and 299 [here](https://github.com/dunhamsj/amrex/blob/development/Src/AmrCore/AMReX_FillPatchUtil.H)
Lines 233, 272, 701, 884, 1276, 1311, 1499, 1527, 1622, 1650 [here](https://github.com/dunhamsj/amrex/blob/development/Src/AmrCore/AMReX_FillPatchUtil_I.H)
Lines 133, 180, 304, 342 [here](https://github.com/dunhamsj/amrex/blob/development/Src/F_Interfaces/AmrCore/AMReX_fillpatch_fi.cpp)
Lines 90, 114, 169, 188, 371, 497, 824, 912 [here](https://github.com/dunhamsj/amrex/blob/development/Src/F_Interfaces/AmrCore/AMReX_fillpatch_mod.F90)

**AverageDown**
Lines 146 and 151 [here](https://github.com/dunhamsj/amrex/blob/development/Src/Base/AMReX_MultiFabUtil.H)
Lines 399, 411, 496, 507 [here](https://github.com/dunhamsj/amrex/blob/development/Src/Base/AMReX_MultiFabUtil.cpp)
Lines 14, 30 [here](https://github.com/dunhamsj/amrex/blob/development/Src/F_Interfaces/Base/AMReX_multifabutil_fi.cpp)
Lines 28, 38, 92, 107 [here](https://github.com/dunhamsj/amrex/blob/development/Src/F_Interfaces/Base/AMReX_multifabutil_mod.F90)

**Reply by WeiqunZhang** [reply]
What's that `CoarseToFineProjectionMatrix`? Do you need to pass it as an argument to FillPatch? Can you just store it inside your interpolater? If you do that, you don't seem to need to modify FillPatch functions at all.

**Reply by dunhamsj** [reply]
`CoarseToFineProjectionMatrix` is a matrix that maps the solution from a coarse reference element to the solution on the fine reference elements. Because of that, it depends on the number of degrees of freedom per element, the dimensionality of the problem, and the choice of basis functions. It is defined once on the user-side at program initialization. It needs to be passed from the user application to amrex, but there could be a separate routine that does that, if it could then be read by the interpolater. I tried to do that at some point, but couldn't get it pass the GPU compilation tests

**Reply by WeiqunZhang** [reply]
> but couldn't get it pass the GPU compilation tests

I suppose there are no reasons that it cannot be done.

No matter where the DG code is, it seems much cleaner to put the matrix into the Interpolater class. Instead of duplicating FillPatch functions, all you need to do is to add a class derived from Interpolater. If you want to put the interpolater into amrex, the matrix initialization code should be included, otherwise it's not that useful for other people.

**Reply by dunhamsj** [reply]
Okay, that all makes sense, thanks. I will try to add a derived class like you mentioned, and I will also try to put in some code to initialize the matrix. For the derived class, how will that be different from what I have at line 1013 [here](https://github.com/dunhamsj/amrex/blob/development/Src/AmrCore/AMReX_Interpolater.H)? Also, do you have a recommendation for where the matrix initialization code be included?

**Reply by WeiqunZhang** [reply]
I would try to add something like
```
enum struct DGBasis { xxx, yyy };
template <int DOF, DGBasis B>
class DGInterpolater
    : public InterpBase
{
public:
    DGInterpolater ()
    {
        initialize matrix here.
    }

private:
   Matrix m_matrix;
};
```

Then somewhere in the user's code, one can define a DGInterplater object with chosen DOF and DG basis function.

**Reply by WeiqunZhang** [reply]
If you get it started, I can help if needed.

**Reply by dunhamsj** [reply]
This is great, thank you! I will work on this and let you know if I get stuck.

**Reply by WeiqunZhang** [reply]
Sounds good. Do interpolation/FillPatch first. Then you can work on average down.

**Reply by dunhamsj** [reply]
Will do, thanks.

**Reply by dunhamsj** [reply]
Thanks to the skeleton you provided, I've defined a class derived from InterpBase and initialized the matrix, and also created `amrex/Tests/DGInterpolater` to test it. I'm confused what to to next though. In particular, how do I access the matrix in the existing FillPatch routines?

**Reply by WeiqunZhang** [reply]
What do you need the matrix for in FillPatch? Passing it to the Interpolater? If so, it already has it. Maybe I am missing something.

**Reply by dunhamsj** [reply]
Ah, I was confused. I didn't have any of the dg_interp functions in my new class. Now it makes sense.

I'm now wondering though how to use my FillPatch routine that takes in four multifabs: two multifabs contain the fluid (one on the coarse level and one on the fine level) and the other two contain the geometry (one for the coarse level and one for the fine level), which is needed to interpolate while conserving the cell averages. Can that be done, or should I create a new multifab that holds the fluid and the geometry?

**Reply by WeiqunZhang** [reply]
I see. In this case, I think you can make your interpolater derive from MFInterpolater instead of Interpolater.
```
DGInterpolater final
    : public MFInterpolater
{
public:
    DGInterpolater(....); // constructor

    Box CoarseBox (Box const& fine, int ratio) override;
    Box CoarseBox (Box const& fine, IntVect const& ratio) override;

    void interp (MultiFab const& crsemf, int ccomp, MultiFab& finemf, int fcomp, int ncomp,
                         IntVect const& ng, Geometry const& cgeom, Geometry const& fgeom,
                         Box const& dest_domain, IntVect const& ratio,
                         Vector<BCRec> const& bcs, int bcomp) override;
}
```
You can then pass additional information into the constructor and store them. Everytime you call FillPatch, you need to make a local DGInterpolater object that has the information you need.

**Reply by dunhamsj** [reply]
I see. I would want this to be something useful to other users and it seems like it's becoming a little too specialized for that. Because of that, I think I'll just keep the additional code on our repo. Thank you very much for your help

**Reply by BenWibking** [reply]
@dunhamsj Just FYI- if it's a consideration as to whether to upstream this, the ability to do fixed-order nodal DG with AMR would indeed be very useful for our code as well.

**Reply by dunhamsj** [reply]
Hi @BenWibking, that's good to know, I was wondering how useful it would be to others. I'm concerned because the way we handle curvilinear coordinates is different from how amrex does it, and I'm worried that may cause confusion. I will keep thinking about it though

**Comment by lwJi** [comment]
Hi @dunhamsj , how did the DG implementation go? Was it successful? I'm really interested ;)
